Release briefing 4.6 (#125)
kokkos/kokkos-tutorials: Content/ReleaseBriefings/4_6/Section_NewFeatures.tex

%    \begin{center}
%        \begin{tabular}{l|l}
%            a & b \\\hline
%            c & d
%        \end{tabular}
%    \end{center}
%\end{frame}

%==========================================================================

% \begin{frame}[fragile]\label{sec:new_features}

  % {\Huge Kokkos::Graph features}

  % \vspace{10pt}

% \end{frame}

\begin{frame}[fragile]{Kokkos::Graph recap}
 \begin{itemize}
     \item describes asynchronous workloads organised as a direct acyclic graph (DAG)
     \item executed using \texttt{submit()}, possibly many times, observing dependencies
      \begin{code}[keywords={auto}]
auto graph = Kokkos::create_graph([&](auto root) {
    auto node_A = root.then_parallel_for("A", ...policy..., ...functor...);
    auto node_B = node_A.then_parallel_for("B", ...policy..., ...functor...);
    auto node_C = node_A.then_parallel_for("C", ...policy..., ...functor...);

    auto node_D = Kokkos::when_all(node_B, node_C).
                  then_parallel_for("D", ...policy..., ...functor...);
});

graph.instantiate();

graph.submit();
      \end{code}
 \end{itemize}
\end{frame}

\begin{frame}[fragile]{Kokkos::Graph new features}
 \begin{itemize}

>   \item \href{https://github.com/kokkos/kokkos/pull/7629}{\texttt{then} node}: executes a callable on device

Reviewer: Fine if you really want them but references to GH issues are not super useful

Author: @romintomasetti

Reviewer 2: This is the PR that introduces the feature. It's a good starting point to learn about it (first version and tests). I always find it useful to get some external references when going through the briefing slides.

Author: I wield my power as person responsible for the release briefing and removed the links in https://github.com/kokkos/kokkos-tutorials/pull/125/commits/c38530db6313a113f719f902c8996c288de47808 (all the other slides are without links to the prs)

   \item Executed in the \texttt{ExecutionSpace} the graph is submitted to
     \begin{code}[keywords={auto}]
auto graph = Kokkos::create_graph([&](auto root) {
    auto node_A = root.then_parallel_for("A", ...policy..., ...functor...);

>     auto node_B = node_A.then(...functor...);

Reviewer: ```suggestion
    auto node_B = node_A.then(KOKKOS_LAMBDA() { Kokkos::printf("hello world from then node\n"); });
```
`...functor...` is not great.  It does not convey anything about the signature and that it differs from then_parallel_for

Author: done in https://github.com/kokkos/kokkos-tutorials/pull/125/commits/2d374769673abb3d1f749581a11a30c782ac0ef1

});

graph.instantiate();

> graph.submit();

Reviewer: I would omit these and use the space to actually say what "then" nodes are about
whether you want to sell that it is a short hand for parallel_for(1, ...) or whatever other reason one might want to use them

Author: I added a bullet point above https://github.com/kokkos/kokkos-tutorials/pull/125/commits/2d374769673abb3d1f749581a11a30c782ac0ef1

     \end{code}
 \end{itemize}
\end{frame}

\begin{frame}[fragile]{Kokkos::Graph new features}
 \begin{itemize}
   \item \href{https://github.com/kokkos/kokkos/pull/7664}{interoperability}: create a \texttt{Kokkos::Graph} from a native Cuda/HIP/Sycl graph

>    \item Does not yet allow linking Kokkos nodes to native nodes

Reviewer: @romintomasetti what does  that mean "linking" nodes

Reviewer 2: It's @JBludau 's wording 😉 But I guess
> It is not yet possible adding nodes created using the native API to a `Kokkos::Graph`

or similar could be an alternative.

Author: it is taken from the pr description in https://github.com/kokkos/kokkos/pull/7664

Reviewer 2: Indeed!

Author: done in https://github.com/kokkos/kokkos-tutorials/pull/125/commits/1dbcacf143a63ad226a68b91dc3c8ad39c18258a

>    \item Does not yet allow linking Kokkos nodes to native nodes

Author: ```suggestion
   \item Does not yet allow adding nodes created using the native API to a \texttt{Kokkos::Graph}
```

>      \begin{code}[keywords={auto}]

Reviewer: There is too much code on the slide, consider highlighting what you want people to see
```suggestion
     \begin{code}[keywords={create_graph_from_native}]
```

cudaGraph_t native_graph = nullptr;

> KOKKOS_IMPL_CUDA_SAFE_CALL(cudaGraphCreate(&native_graph, 0));

Reviewer: No KOKKOS_IMPL on slides

Author: done in https://github.com/kokkos/kokkos-tutorials/pull/125/commits/7896362a42f14ff6e2355db42ba0540e4a2e10c3

auto graph_from_native =
  Kokkos::Experimental::create_graph_from_native(exec, native_graph);

> auto root = Kokkos::Impl::GraphAccess::create_root_ref(graph_from_native);

Reviewer: No Kokkos::Impl::

Author: hmmm ... looks like I can not find any public interface to get the root via a public interface @romintomasetti

Author: should we remove this for now?

root.then_parallel_for(1, Increment<view_t>{data});
...
graph_from_native.submit(exec);
     \end{code}
 \end{itemize}
\end{frame}

% \begin{frame}[fragile]\label{sec:new_features}

  % {\Huge Multi-GPU for HIP Backend}

  % \vspace{10pt}

% \end{frame}

\begin{frame}[fragile]{Multi-GPU for HIP Backend}
  \begin{itemize}
    \item Launch kernels on multiple devices from a single host process
    \item Available for ROCm 5.6 and later

>     \item Requires knowledge of HIP API functions for creating and destroying streams

Reviewer: knowledge?
change to "direct use of HIP runtime API"?

Reviewer: You might want to mention setting the device on which the stream will be created and manage the stream lifetime

Reviewer 3: Updated. I put setting device and creating/destroying streams in the example. I may just mention verbally then.

>     \item Experimental, always looking for feedback from new users

Reviewer: ```suggestion
    \item Experimental, still looking for feedback from new users
```

\begin{frame}[fragile]{Kokkos::Graph recap}
 \begin{itemize}
     \item describes asynchronous workloads organised as a direct acyclic graph (DAG)
     \item executed using \texttt{submit()}, possibly many times, observing dependencies
      \begin{code}[keywords={auto}]
auto graph = Kokkos::create_graph([&](auto root) {
    auto node_A = root.then_parallel_for("A", ...policy..., ...functor...);
    auto node_B = node_A.then_parallel_for("B", ...policy..., ...functor...);
    auto node_C = node_A.then_parallel_for("C", ...policy..., ...functor...);

    auto node_D = Kokkos::when_all(node_B, node_C).
                  then_parallel_for("D", ...policy..., ...functor...);
});

graph.instantiate();

graph.submit();
      \end{code}
 \end{itemize}
\end{frame}

\begin{frame}[fragile]{Kokkos::Graph new features}
 \begin{itemize}
  \item \texttt{then} node: executes a callable on device
  \item Single call of the functor per \texttt{submit()}
   \item Executed in the \texttt{ExecutionSpace} the graph is submitted to
     \begin{code}[keywords={auto}]
auto graph = Kokkos::create_graph([&](auto root) {
    auto node_A = root.then_parallel_for("A", ...policy..., ...functor...);

>     auto node_B = node_A.then(KOKKOS_LAMBDA(){ Kokkos::printf("hello\n");});

Author: It already is `Kokkos::printf` ... I tink it is ok for illustrative purposes.
Also, the second bullet point is already spelling out that it is executed in a given Space.

...
graph_from_native.submit(exec);
     \end{code}
 \end{itemize}
\end{frame}

% \begin{frame}[fragile]\label{sec:new_features}

  % {\Huge Multi-GPU for HIP Backend}

  % \vspace{10pt}

% \end{frame}

\begin{frame}[fragile]{Multi-GPU for HIP Backend}
  \begin{itemize}
    \item Launch kernels on multiple devices from a single host process
    \item Available for ROCm 5.6 and later
    \item Requires direct use of HIP runtime API for creating and destroying streams
    \item Experimental, still looking for feedback from new users
    \item Newly documented (for all backends)  
      \begin{itemize} 
        \item[] \url{https://kokkos.org/kokkos-core-wiki/API/core/MultiGPUSupport.html}
      \end{itemize}
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{Multi-GPU for HIP Backend}
  \begin{code}[keywords={auto}]
// Create streams on different devices
hipStream_t streams[2];
hipSetDevice(0); hipStreamCreate(&streams[0]);
hipSetDevice(1); hipStreamCreate(&streams[1]);
{
  // Creating execution spaces 
  Kokkos::HIP exec0(streams[0]), exec1(streams[1]);

  // Allocating views
  Kokkos::View<int*> v0(Kokkos::view_alloc("v0", exec0), N);

>   Kokkos::View<int*> v1(Kokkos::view_alloc("v1", exec0), M);

Reviewer 3: ```suggestion
  Kokkos::View<int*> v1(Kokkos::view_alloc("v1", exec1), M);
```

%==========================================================================

\begin{frame}[fragile]\label{sec:new_features}

  {\Huge Kokkos::Graph features}

  \vspace{10pt}

\end{frame}

\begin{frame}[fragile]{Kokkos::Graph recap}
 \begin{itemize}
     \item Allows creating a graph of kernels

>      \item Executed in order when calling \texttt{submit(ExecutionSpace)}

Reviewer 2: Let's reuse the same wording as in https://github.com/kokkos/kokkos-core-wiki/pull/648.
```tex
\item describes asynchronous workloads organised as a direct acyclic graph (DAG)
\item executed using \texttt{submit()}, possibly many times, observing dependencies
```
(I've removed the execution space here too)

>      \item Allows creating a graph of kernels
>      \item Executed in order when calling \texttt{submit(ExecutionSpace)}

Author: ```suggestion
     \item describes asynchronous workloads organised as a direct acyclic graph (DAG)
     \item executed using \texttt{submit()}, possibly many times, observing dependencies
```

      \begin{code}[keywords={auto}]
    ...
    auto graph = Kokkos::Experimental::create_graph(exec, [&](auto root) {
      auto node_A = root.then_parallel_for(
                    policy_t(exec, 0, 1),
                    FetchValuesAndContribute(data, index_A, value_A));
      auto node_B = root.then_parallel_for(
                    policy_t(exec, 0, 1),
                    FetchValuesAndContribute(data, index_B, value_B));}
    ...

>     graph.submit(exec);

Reviewer 2: Please make it more appealing, and aligned with the example I gave in https://github.com/kokkos/kokkos-core-wiki/pull/648. And remove `exec` because it may be confusing and isn't serving any particular purpose in this case.
```c++
    auto graph = Kokkos::create_graph([&](auto root) {
        auto node_A = root.then_parallel_for("workload A", ...policy..., ...functor...);

        auto node_B = node_A.then_parallel_for("workload B", ...policy..., ...functor...);
        auto node_C = node_A.then_parallel_for("workload C", ...policy..., ...functor...);

        auto node_D = Kokkos::when_all(node_B, node_C).then_parallel_for("workload D", ...policy..., ...functor...);
    });

    graph.instantiate();

    graph.submit()
```

Author: done in https://github.com/kokkos/kokkos-tutorials/pull/125/commits/ba8f0b308d21e5508c5a7d60179eaa58717f799d

     \end{code}
 \end{itemize}
\end{frame}

\begin{frame}[fragile]{Kokkos::Graph new features}
 \begin{itemize}

>    \item \texttt{then}-node: adds callable to be executed once

Reviewer 2: "to be executed once" might be misleading with the fact that the graph can be submitted many times. I'm not sure my proposal is better, but you get the point :wink:

```tex
\item \href{https://github.com/kokkos/kokkos/pull/7629}{\texttt{then} node}: executes a callable on device
```

>    \item \texttt{then}-node: adds callable to be executed once

Author: ```suggestion
  \item \href{https://github.com/kokkos/kokkos/pull/7629}{\texttt{then} node}: executes a callable on device
```

   \item Executed in the \texttt{ExecutionSpace} the graph is submitted to
     \begin{code}[keywords={auto}]
    ...
    auto graph = Kokkos::Experimental::create_graph(exec, [&](auto root) {
      auto node_A = root.then_parallel_for(
                    policy_t(exec, 0, 1),
                    FetchValuesAndContribute(data, index_A, value_A));
      auto node_B = node_A.then(
                    FetchValuesAndContribute(data, index_B, value_B));}
    ...
    graph.submit(exec);

>      \end{code}

Reviewer 2: I thought this was part of the `4.5` release ? https://github.com/kokkos/kokkos-tutorials/blob/21c0ccb69716deb7acabc052857089071116df43/Content/ReleaseBriefings/4_5/Section_GraphsEnhancements.tex#L75

Author: yes, I reiterated it here so there is no confusion like "oh, the then nodes are executed on the host?"

 \end{itemize}
\end{frame}

\begin{frame}[fragile]{Kokkos::Graph new features}
 \begin{itemize}

>    \item Interoperabiliy: create a Kokkos::Graph from CUDA/HIP/SYCL graph

Reviewer 2: ```tex
\item \href{https://github.com/kokkos/kokkos/pull/7664}{interoperability}: create a \texttt{Kokkos::Graph} from a native Cuda/HIP/Sycl graph
```

>    \item Interoperabiliy: create a Kokkos::Graph from CUDA/HIP/SYCL graph

Author: ```suggestion
   \item \href{https://github.com/kokkos/kokkos/pull/7664}{interoperability}: create a \texttt{Kokkos::Graph} from a native Cuda/HIP/Sycl graph
```
